File: Sources/App/Actions.swift
import AppKit

enum Actions {
    static func openInBrowser(port: Int) {
        if let url = URL(string: "http://localhost:\(port)") {
            NSWorkspace.shared.open(url)
        }
    }

    static func openURL(_ string: String) {
        if let url = URL(string: string) { NSWorkspace.shared.open(url) }
    }

    static func copy(_ text: String) {
        NSPasteboard.general.clearContents()
        NSPasteboard.general.setString(text, forType: .string)
    }

    /// SIGTERM the process. Returns false if the call fails.
    @discardableResult
    static func kill(pid: Int32) -> Bool {
        Foundation.kill(pid, SIGTERM) == 0
    }
}

File: linux/porthole/src/actions.rs
//! Port actions. Linux counterpart of `App/Actions.swift`.

use gtk::prelude::*;

pub fn open_in_browser(port: u16) {
    open_url(&format!("http://localhost:{port}"));
}

pub fn open_url(url: &str) {
    // `xdg-open` picks the user's default browser (via the portal under Flatpak);
    // ignore failure (no browser installed).
    let _ = std::process::Command::new("xdg-open").arg(url).spawn();
}

pub fn copy(text: &str) {
    if let Some(display) = gtk::gdk::Display::default() {
        display.clipboard().set_text(text);
    }
}

/// SIGTERM the process. Returns false if the call fails.
pub fn kill(pid: i32) -> bool {
    // Safe: kill(2) with SIGTERM has no memory effects.
    unsafe { libc::kill(pid, libc::SIGTERM) == 0 }
}
